### ISSUE TITLE
Subnormals and other performance issues

### ISSUE THREAD

**Issue by MikhailRyazanov** [PyAbel/PyAbel#246]
This is a continuation of the discussion from #244 about the strange benchmark results obtained on Intel CPUs.

Basically, the observation was that on various Intel CPUs `basex` and `basex(var.reg.)` do the forward transform much slower than the inverse transform (https://github.com/PyAbel/PyAbel/pull/244#issuecomment-468479433), whereas on an AMD CPU they have the equal speeds (https://github.com/PyAbel/PyAbel/pull/244#issuecomment-468353490), as expected.

This is apparently caused by a significant penalty of working with [subnormal floating-point numbers](https://en.wikipedia.org/wiki/Denormal_number) that Intel CPUs have (and at least AMD Ryzen does not).

The “proper” way to avoid this penalty, recommended by Intel is to [set the related CPU flags](https://software.intel.com/en-us/mkl-linux-developer-guide-operating-on-denormals) such that subnormals are treated as zeros. The problem is that neither NumPy (numpy/numpy#4581), nor the underlying matrix libraries (OpenBLAS (xianyi/OpenBLAS#1237), Intel MKL, Apple's Accelerate Framework) expose this capability.

There is, however, a proof-of-concept module [daz](https://github.com/chainer/daz), which does just that (sets the CPU flags) from Python. It works as expected with MKL (see also [matrix benchmarks](https://gist.github.com/MikhailRyazanov/63823e1bc7028085389d787b765a4f2c)):
```
Python 2.7.15 |Anaconda, Inc.| (default, Dec 14 2018, 19:04:19) 
[GCC 7.3.0] on linux2
using Intel MKL

=== Basis generation ==============================   with daz:

Method              n = 513    n = 1025    n = 2049   n = 513    n = 1025    n = 2049
---------------------------------------------------   -------------------------------
basex                1486.2      9070.2     43671.3    1452.8      8901.9     43356.8
linbasex               19.8        74.2       269.9      19.8        68.7       280.1
onion_peeling           4.1        18.6       101.7       4.2        18.1        97.2
three_point            14.4        50.7       212.7      14.4        49.5       205.1
two_point               3.2        12.7        50.7       3.2        12.8        51.8


=== Forward Abel transform ========================   with daz:

Method              n = 513    n = 1025    n = 2049   n = 513    n = 1025    n = 2049
---------------------------------------------------   -------------------------------
basex                   1.2        50.5       340.1       1.2         8.6        64.9
basex(var.reg.)        10.8       135.1       623.2       7.9        42.1       268.0
direct_Python         480.0      4239.3     33798.4     482.6      4117.0     34233.3
hansenlaw              28.5       116.1       439.0      28.7       111.1       422.4


=== Inverse Abel transform ========================   with daz:

Method              n = 513    n = 1025    n = 2049   n = 513    n = 1025    n = 2049
---------------------------------------------------   -------------------------------
basex                   1.2         9.6        68.1       1.2         8.4        63.2
basex(var.reg.)        10.3        61.5       318.9       8.0        42.5       275.7
direct_Python         483.5      4236.4     34387.9     484.3      4092.0     34299.5
hansenlaw              29.3       119.0       446.2      29.4       112.7       430.9
linbasex              172.5       949.1      5578.7     173.0       938.9      5672.9
onion_bordas          324.1      2777.0     11498.5     341.9      1897.8     11774.1
onion_peeling           1.2         9.0        69.8       1.4         8.4        61.5
three_point             1.2         8.6        63.7       1.2         8.2        59.7
two_point               1.2         8.6        64.4       1.3         8.4        59.5
```
and seems to work with OpenBLAS:
```
Python 2.7.15+ (default, Feb  3 2019, 13:13:16) 
[GCC 8.2.0] on linux2
using OpenBLAS

=== Basis generation ==============================   with daz:

Method              n = 513    n = 1025    n = 2049   n = 513    n = 1025    n = 2049
---------------------------------------------------   -------------------------------
basex                1319.3      8675.8     62820.0    1350.6      8785.3     63229.6
linbasex               20.3        73.5       294.0      19.7        74.0       289.7
onion_peeling           4.1        17.3       102.2       4.0        17.6        91.0
three_point             9.6        41.3       182.5       9.5        42.0       180.7
two_point               3.2        11.2        47.3       3.2        11.1        46.5


=== Forward Abel transform ========================   with daz:

Method              n = 513    n = 1025    n = 2049   n = 513    n = 1025    n = 2049
---------------------------------------------------   -------------------------------
basex                   1.4        50.1       295.9       1.4        11.6        79.1
basex(var.reg.)        11.1       105.7       625.1       7.0        41.5       248.3
direct_Python         364.8      3432.3     28515.2     365.1      3490.3     28167.9
hansenlaw              27.1       106.5       408.0      27.4       106.4       407.3


=== Inverse Abel transform ========================   with daz:

Method              n = 513    n = 1025    n = 2049   n = 513    n = 1025    n = 2049
---------------------------------------------------   -------------------------------
basex                   1.4        11.1        80.8       1.4        11.0        72.0
basex(var.reg.)         9.9        57.0       344.2       7.4        45.6       256.1
direct_Python         381.4      3449.7     28683.6     380.5      3508.1     28806.9
hansenlaw              27.7       107.7       413.5      28.0       109.3       412.9
linbasex              224.1      1055.4      6632.8     214.9      1044.8      6430.2
onion_bordas          301.8      1696.4     12099.7     309.5      1709.3     12283.6
onion_peeling           1.4        10.8        89.1       1.4        10.6        75.8
three_point             1.3        10.1        74.4       1.3        10.1        73.7
two_point               1.3        10.5        78.1       1.3        10.2        69.9
```
but matrix tests show that for OpenBLAS ([see here](https://gist.github.com/MikhailRyazanov/63823e1bc7028085389d787b765a4f2c#file-output-intel-core-i5-4278u-2-60-ghz-debian-python-openblas-daz)) the problem is not completely corrected and for Apple's Accelerate Framework ([see here](https://github.com/PyAbel/PyAbel/pull/244#issuecomment-472564531)) is not corrected at all(?).

Another issue is that at least in one case a “high-end” Intel CPU [shows](https://github.com/PyAbel/PyAbel/pull/244#issuecomment-471891926) very poor matrix-inversion performance, both with and without subnormals.

**Comment by MikhailRyazanov** [comment]
One more observation from the tables above is that Anaconda seems to be slower than the standard Debian Python in cases when the interpreter itself is expected to do more work (`direct_Python` and smaller-size `onion_bordas`). Does Anaconda have a poorly optimized Python?

This is Python 2.7 in both cases. I can also try to test Anaconda Python 3, but my Debian Pythons 2.7 and 3.7 show no significant differences.

**Comment by MikhailRyazanov** [comment]
Regarding subnormals, we could just recommend those with Intel CPUs to use MKL (which is anyway the fastest, although by only ~10% on my system) and the `daz` module, and those with AMD CPUs should not have this problem at all.

But MKL is not a standard option for NumPy (at least I did not find how to marry them under Debian without recompiling manually, and even just installing MKL itself tries to pull crazy gigabytes of some junk...).

`daz`, although very simple, is no more than somebody's experiment. And requires a C compiler for installation.

I think, we should at least try to persuade NumPy to incorporate the `daz` functionality. And then attempt to encourage OpenBLAS to honor the user-set DAZ and FTZ flags. Apple is probably hopeless in this respect.

**Comment by MikhailRyazanov** [comment]
And as a workaround that does not depend on other people, if small numbers are manually removed from all matrices (this is what [bm-mat.py](https://gist.github.com/MikhailRyazanov/63823e1bc7028085389d787b765a4f2c#file-bm-mat-py) does for `Xc` and `Yc`), then the results (ignore the `X` and `Y` rows and columns in the tables) look not so bad. I've tried this with `basex`, and it does help significantly. By it somewhat clutters the code, so I do not like this solution...

**Comment by DanHickstein** [comment]
This is fascinating from the computer science aspect, and really neat in our efforts to eek out every bit of performance from our algorithms. But, it's probably not relevant to 99% of PyAbel users, and the solutions basically amount to us cluttering our code to work-around what is basically a bug/limitation in NumPy (as far as I understand).

> I think, we should at least try to persuade NumPy to incorporate the daz functionality. And then attempt to encourage OpenBLAS to honor the user-set DAZ and FTZ flags. 

Yeah, this seems like something that NumPy should at least consider. Is this the most recent discussion? https://github.com/numpy/numpy/issues/4581  I'd say that you either revive this thread or start a new one.

One solution is that we could incorporate the daz code into PyAbel (it's MIT license). But, it seems excessive to have another c-compiled part of our code just for the sake of dealing with denormals.

For the benchmarks, this does seem really important, and we could have some sort of try/except block that uses `daz` if it's installed.

**Comment by MikhailRyazanov** [comment]
Yes, I do not want to clutter our code or make it C-dependent. Using `daz` automatically is also not the best option, since AMD does not need it, and it actually makes the calculations slightly less accurate and break the IEEE 754 standard (although both these issues are not really important).

I think, we should just add a section about performance recommendations to the docs. And there suggest to use `daz` if needed.

And I'll try to see what NumPy and OpenBLAS can do about this...

**Comment by MikhailRyazanov** [comment]
There one more strange thing that needs investigation: [Subnormals appears to have an important effect on real images, even for the inverse Abel transform](https://github.com/PyAbel/PyAbel/pull/244#issuecomment-472586409).

**Comment by DanHickstein** [comment]
> I think, we should just add a section about performance recommendations to the docs.

Yeah, I think that this is probably the way to go. We can make add a line in the benchmarks function and the Transform function sending users to a new doc page about performance considerations.

**Comment by MikhailRyazanov** [comment]
I'm going to close this, since our documentation now (after #283) at least warns the users about these issues and gives some advice, but we cannot do much more in PyAbel...
